pkg/applicationserver/io/mqtt/mqtt.go

```diff
@@ -129,31 +129,31 @@ func (c *connection) setup(ctx context.Context) error {
 				var topicParts []string
 				switch up.Up.(type) {
 				case *ttnpb.ApplicationUp_UplinkMessage:
-					topicParts = c.format.UplinkTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.UplinkTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_JoinAccept:
-					topicParts = c.format.JoinAcceptTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.JoinAcceptTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_DownlinkAck:
-					topicParts = c.format.DownlinkAckTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.DownlinkAckTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_DownlinkNack:
-					topicParts = c.format.DownlinkNackTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.DownlinkNackTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_DownlinkSent:
-					topicParts = c.format.DownlinkSentTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.DownlinkSentTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_DownlinkFailed:
-					topicParts = c.format.DownlinkFailedTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.DownlinkFailedTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_DownlinkQueued:
-					topicParts = c.format.DownlinkQueuedTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.DownlinkQueuedTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				case *ttnpb.ApplicationUp_LocationSolved:
-					topicParts = c.format.LocationSolvedTopic(unique.ID(c.io.Context(), c.io.ApplicationIDs()), up.DeviceID)
+					topicParts = c.format.LocationSolvedTopic(unique.ID(up.Context, c.io.ApplicationIDs()), up.DeviceID)
 				}
 				if topicParts == nil {
 					continue
 				}
-				buf, err := c.format.FromUp(up)
+				buf, err := c.format.FromUp(up.ApplicationUp)
 				if err != nil {
 					logger.WithError(err).Warn("Failed to marshal upstream message")
 					continue
 				}
-				logger.Info("Publish upstream message")
+				logger.WithField("device_uid", unique.ID(up.Context, up.EndDeviceIdentifiers)).Debug("Publish upstream message")
```

Comment:
Can you add the field to the logger once you have the contextual up? This is going to be interesting in other log entries as well